draft-irtf-cfrg-opaque.md

```diff
@@ -398,145 +398,331 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+Steps:
+1. prk = Extract(envelope.nonce, Harden(y, params))
+2. auth_key = Expand(prk, "AuthKey", Nx)
+3. export_key = Expand(prk, "ExportKey", Nx)
+4. cleartext_creds = CreateCleartextCredentials(response.server_public_key, client_public_key, creds.server_identity, creds.client_identity)
+5. expected_tag = MAC(auth_key, concat(mode, envelope_nonce, inner_env, cleartext_creds))
+6. If !ct_equal(envelope.auth_tag, expected_tag),
+    raise MacError
+7. secret_creds = RecoverSecret(prk, envelope.inner_env)
+8. Output (secret_creds, export_key)
+~~~
 
-These credentials are incorporated in the `SecretCredentials` and `CleartextCredentials` structs,
-depending on the mode set by the value of `EnvelopeMode`:
+## Envelope Modes {#envelope-modes}
 
+The `EnvelopeMode` specifies the structure and encoding of the corresponding `InnerEnvelope`. This document specifies the values of the two aformentionned modes:
 ~~~
 enum {
-  base(1),
-  custom_identifier(2),
-  (255)
+ internal(1),
+ external(2),
+ (255)
 } EnvelopeMode;
 ~~~
 
-The `base` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+Each `EnvelopeMode` defines its own `InnerEnvelope` structure and must implement the following interface:
+- `inner_env, client_public_key = BuildInnerEnvelope(prk, creds)`: build and return the `InnerEnvelope` structure and the client's public key.
+- `secret_creds = RecoverSecret(prk, inner_env)`: recover and return the `SecretCredentials` structure.
+
+The inputs to Extract and Expand are as specified in {{RFC5869}}. The underlying hash function
+is that which is associated with the OPAQUE configuration (see {{configurations}}).
+
+### Internal Key mode {#internal-mode}
+
+In this mode, the client's private and public keys are deterministically derived from the OPRF output. For OPAQUE-3DH, this is done by hashing the OPRF output to a scalar in the corresponding prime order group to produce a private key, and then multiplying that scalar by the fixed generator `G` to produce the public key.
 
+With the internal key mode the `EnvelopeMode` value MUST be `internal` and the `InnerEnvelope` is empty of nil. Indeed, this mode does not need it.
 ~~~
-struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+BuildInnerEnvelope(prk, creds)
 
-struct {
-  opaque server_public_key[Npk];
-} CleartextCredentials;
+Parameters:
+- group, the AKE group to derive the keys in
+
+Input:
+- prk, pseudo-random key
+- creds, Credentials structure. Not used in this function, it only serves to comply with the API
+
+Output:
+- inner_env, nil value (serves to comply with the API)
+- pkc, the client's AKE public key
+
+Steps:
+1. _, pkc = DeriveKeyPair(prk)
+2. Output (nil, pkc)
 ~~~
 
-The `custom_identifier` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+OPAQUE-3DH uses HashToScalar(seed) to derive the client's secret key in this mode.
+~~~
+DeriveKeyPair(prk)
```

Comment:
I think we should move this to the configuration section, which discusses particular types of Groups for 3DH.

Reply:
Moved to the configurations sections, please have a look